### ISSUE TITLE
C++ interface for GAP/LibGAP

### ISSUE THREAD

**Issue by innodocs** [gap-system/gap#4375]
Hello.

as I needed some math functionality for one of my C++ projects, I looked at GAP/libGAP. it is indeed working AS IS, but a bit of a learning curve for developers new to the system.

To make the work of other devs on the team easier, I added a C++ wrapper over LibGAP, see [C++ LibGAP Interface](https://github.com/innodocs/GAP).

Only integers and rationals, and stack objects thus far. I'll add permutations, lists, vectors and matrices as we need them, and also support for including/referencing GAP GC objects from C++ heap objects.

In case you are interested in this addition to GAP, there are a number of changes needed to the GAP kernel, see the 'Building the System' section in the [Readme](https://github.com/innodocs/GAP/blob/master/README.md) file. Most are in the _"remove static from function definition"_ category, but I foresee three bigger changes needed down the road:

- integrating the C++ exception handling mechanism w/ the GAP error handling
- streams to capture, re-direct, manipulate GAP output in C++, w/o having to buffer the output via a string (toString() call). for large outputs toString will be slow/inefficient/impracticable
- a more efficient interface tho the GC than MarkBag, allowing bulk marking and a very fast (inlined) test to determine whether a pointer could be a GAP GC object

Thank you

**Comment by ChrisJefferson** [comment]
Hi,

It's always nice to see more progress in this area. I have my own (very small) similar library at: https://github.com/ChrisJefferson/gapcpp

Looking at your final big three issues, my belief (without having tried them in depth)

1) Making exception handling work with GAP error handling is, I believe, possible but will be significant amounts of work. This area needs to be edited very careful

2) Generally improving how GAP's input and output streams work has been a multi-year, and continuing project. I would look into how both Sage's GAP interface, and GAP.jl (the Julia interface) work, to see the current best way of doing this.

3) It would be good to know why exactly you need this -- is it connected to putting GAP items into C++ containers? My initial thought is if you need to know so quickly what is a GAP object, and bulk marking, you are (probably) "doing it wrong", you should only mark things when they change (and the interface tries to do that itself wherever possible). Storing GAP objects efficiently inside C++ containers is very difficult, past interfaces with significant chunks of C++ code (like the 'semigroups' and 'ferret' packages) convert into "pure" C++ objects early in the process, both because this is more efficiencent, and it avoids this problem.

**Comment by innodocs** [comment]
1. there are examples in the [test](https://github.com/innodocs/GAP/tree/master/libgap/test) folder on uses of the system. it is surprisingly clean (and elegant) C++ - IMHO.

2. I am using the pattern in [gap/exception.h](https://github.com/innodocs/GAP/blob/master/libgap/include/gap/exception.h) when defining exception hierarchies, and it works pretty well. not sure if that's something which would work for the kernel.

3. I will keep adding to this current system, but then, once I have all the GAP types I need (int, rat, perm, vector, matrix), I plan to call it v1, and then for v2, wrap the class definitions into something like

        namespace GAP::T {
            template<typename GObj, typename GObj_const>
            class Int {...}
        }

and instantiate w/

     namespace GAP::S {
       using Int = T::Int<GAP_Obj, GAP_Obj>
      ...
     }
  
     namespace GAP::H {
       using Int = T::Int<shared_ptr<GAP_Obj>, const shared_ptr<GAP_Obj>>
      ...
     }

'S' - stack
'H' - heap

w/ properly written copy, move operators and conversion S <-> H, this should be fairly efficient, but I'll have to test it (against v1).

Thanks

**Comment by fingolfin** [comment]
Some remarks:

First off, it's great to hear that our work on libgap is of some use to you, and we are happy to discuss ways to improve it. Thank you also for taking the time to write to us, I'll try to have a closer look at your code later!

However, it is currently a non-goal to make this "easy to use for the new developer". Our main product is GAP, the interactive system. We made an API available for some expert developers who requested it. We'll continue to improve it within the strongly limited bounds of our available resources (mostly work is provided by volunteers), and our own needs. But if anyone wants to embed GAP, it's up to them to figure out a lot about it. So e.g. we won't write extensive documentation about it (I guess if someone were to submit such documentation and it has an acceptable quality level, we'd 
probably merge it).

But whenever we add something, it comes at a cost: we must maintain it. And we must thus be *able* to maintain it. We are not using some *very* limited subset of C++ in the kernel, and that is intentional, and not going to change for the time being. This is not because we don't appreciate your needs, or are unwilling to help.

So in particular, I don't think we should *use* C++ exceptions in GAP, at least at this time, as they can result in serious integration hurdles. That said, I am happy to add code which makes it easier to integrate GAP error handling and C++ exceptions. I won't work on it myself, but I'll be happy to review PRs and give advice. I do plan to continue working on improvements to the GAP error handling, though, and e.g. make them integrate better with Julia exceptions, and there is a good chance this will also help C++ (and Python, and others).

On "remove static from function definition", that means you are implicitly trying to make these functions are part of the libgap API. But that's not the way we currently mean to play this game: really, only functions in `src/libgap-api.h` are part of the official API. If you want to use anything else, you are leaving the territory of things about which we make any promises, and it may crash and burn with the next GAP release (not intentionally, of course). This is quite important to me, as we don't want to put ourselves into a situation where we can't make a refactoring because it'd break a promise we made to somebody outside, in code that we can't modify. 

There is a reason why functions like `NUM_RAT` and `SumRat` and what not are not public: they perform absolutely no error handling, and will cause crashes or worse when called with incorrect arguments. Our code is designed carefully to only call them in situations where this is safe. From a quick look at your code, it seems it is "safe" in that your `Rat` class does not actually allow using it for existing objects -- which makes me wonder how uesful it is; if I call a GAP function returning a rational, there seems to be no way to turn it into a C++ `Rat`?

In almost all cases, it is better to invoke high level helpers like `SUM` to add two rationals (mind you, that also isn't part of the official libgap API -- but it's much more likely we'll add that than other stuff. Of course it comes with a performance impact, but I'd suggest to first benchmark realistic code (and not just some micro benchmarks) to see whether it really matters in practice.

**Comment by innodocs** [comment]
thanks so much for taking the time to respond in such detail.

I just posted this so you are aware of it. as I mentioned, we need some math in one of our projects, and instead of re-inventing the wheel I just thought we could use GAP - but wasn't even aware that you had a C library, so that came as a great/welcome surprise. made the whole set-up a breeze.

we need a few more types (permutations, vectors, matrices) so I'll add those as time permits and as needed.

I haven't yet gotten around to test the error handling out, but was planning to experiment w/ the `errorCallback` in `GAP_Initialize`, maybe it's more than enough.

as to:

>if I call a GAP function returning a rational, there seems to be no way to turn it into a C++ Rat?

my intention was to use the C++ type system for insuring type (argument/return value) safety. you can type-cast a GAP handle in a controlled way, for example

        inline Int& Int::operator+=(const Int& opR)
        {
          *this = Int(GAP_SumInt(gapObj, opR.gapObj));
          return *this;
        }

the `Int` explicit constructor will create a C++ `Int` object from the GAP handle returned by `SumInt`. For object creation outside it's own class (e.g. creating `Int`s in a `Rat` method) I'm using a template `apply` function:

        inline Int Rat::mod(const Int& opR) const
        {
          return apply<Int>(ModRat(gapObj, unapply(opR)));
        }
        
this (appears to) work fine - as long as GAP does "typed" returns, or maybe I should say predictable ones, i.e. does not return an Int under some conditions and a Rat at other times.

**Comment by innodocs** [comment]
forgot to mention one more issue, maybe important in understanding how we plan to use/not break/ the system: I don't plan to use any of the GAP kernel functionality for reading in code written in the GAP language, evaluating it, aso asf

if we need GAP functions outside of what's provided in the kernel, we'll use a small translator from GAP to C++, which will generate code based on these base C++ classes.

so, in essence, the GAP kernel for our purposes will be the base system, the garbage collector, and the basic operations on integers, rationals, but no REPL.

that should reduce the chances that we'll be running into unforeseen conditions.

**Comment by fingolfin** [comment]
If those are your needs, using GAP may be total overkill. Have you considered just using GMP directly (it has a nice C++ wrapper), plus permlib for permutation groups, plus one of the many libraries for linear algebra?

Not that I want to dissuade you from using GAP for this, it just seems like a somewhat odd choice if you are never calling into the GAP userspace.

on another note, you wrote
>  as long as GAP does "typed" returns, or maybe I should say predictable ones, i.e. does not return an Int under some conditions and a Rat at other times.

GAP does that all the time because e.g. 1/2 + 1/2 = 1 is an integer, even though the input were two rationals. This is important and means your code is in fact buggy because it calls MakeRat directly which does not sanitize its input and returns denotmalized fractions such as 1/1 (instead of the integer 1), or 2/4 or even 4/2. This may even work for some operations but all kind of things can break. The correct way to construct a fraction is by calling QUO (QuoRat if you are sure the inputs are rationals/integers).

**Comment by fingolfin** [comment]
Regarding my point about not being able to construct a Rat out of any Obj: the relevant constructor is protected, which strikes me as impractical if you ever want to call a function which returns a GAP integer but which isn't itself about integers - so it doesn't really belong into class Int.

But if you make it public, then it should IMHO validate the type of its input (via TNUM_OBJ)

**Comment by innodocs** [comment]
>Regarding my point about not being able to construct a Rat out of any Obj: the relevant constructor is protected, which >strikes me as impractical if you ever want to call a function which returns a GAP integer but which isn't itself about integers - >so it doesn't really belong into class Int.

it would be unsafe for that (explicit) constructor to be public. any method in the class hierarchy can create  an `Int` from a GAP handle via `apply<Int>(gapHandle)`

for other functions, they can either create `Ints` from C-ints, e.g.:

      Gap::Int sumOfSeries(unsigned long start, unsigned long end) {
          if (start > end)
            return 0;
      
          Gap::Int diff = end - start;
          return (diff % 2 == 0) // or simply use diff.isEven()
             ?  (diff+1)    * ((start+end)/2)
             : ((diff+1)/2) *  (start+end);
        }

or as results of other functions. don't see why someone would want to create a C++ object from a GAP handle directly. how would they even get such a handle?

a few comments:

1. because you asked before whether I checked performance: in [libgap/test/Readme](https://github.com/innodocs/GAP/tree/master/libgap/testl) there are some timings, e.g. for `PE-001-gap.cpp` with and without inlined immediate int operations. significant difference.

2. using a bunch of other libs and trying to get them to work together: did an eval, and seems like more work. w/ GAP we know that everything from ints to matrices work fine together. not in the mood to chase after crazy bugs across 5 different libraries.

3.
>overkill ... if you are never calling into the GAP userspace.

plan to use GAP userspace functions, just not through the REPL - we'll precompile the ones we need.

4. can't use `Int` w/o the namespace identifier - because of the clash w/ the kernel `Int`. we'll just have to live w/ that, I guess.

**Comment by ChrisJefferson** [comment]
Many functions in GAP can return multiple types, in particular, many functions return `fail` to denote they failed to find a result -- you could of course generically map such values to an `optional<T>`. 

I would definitely add an always on check (could be a throw, or an assert) that you check you only construct Int objects from a GAP int -- personally I think you *need* to have checks whenever you convert a GAP object to a C++ object, else you are going to get many bugs and incorrect results which are basically impossible to track down.

**Comment by innodocs** [comment]
`integer.h` :

      Obj SignInt( Obj op )
      {
        ...
        return Fail;
      }
      
`int.h`:
 
      inline int Int::sign() const
      {
        GAP_Obj sign = SignInt(gapObj);
        if      (sign == INTOBJ_INT(0))   return 0;
        else if (sign == INTOBJ_INT(1))   return 1;
        else if (sign == INTOBJ_INT(-1))  return -1;
        else /* Fail */
          throw FailedOpException("Int::sign(): bad sign");
      }

**Comment by fingolfin** [comment]
@innodocs not sure what the point of that last one is? Yes, `SignInt` is safe to call on arbitrary objects. But e.g. `AInvInt` is not, nor are many others. I already explained that `MakeInt` is used incorrectly; as a result, I'd expect `Rat(1) == 2*Rat(1,2)` to return `false` -- though of course I might be missing someth

Similarly, I'd expect this to crash or (worse) produce garbage:  `den(2*Rat(1,2))`

Another note: in you `gap-system.h` you have this comment regarding using `GAP_Enter`/`GAP_Leave`
```
**  This is not needed in most cases as the C and C++ stacks are one and the
**  same. If you need this support in your application, define 'SYS_MARK_TOS'
```
This is potentially misleading and IMHO a bit cavalier. E.g. if you write code that calls your `Init` wrapper around `GAP_Initialize` and then at the same level creates GAP objects on the stack, those have a chance of *not* being seen by the GAP garbage collector, as it will have an incorrect value for `StackBottomBags` on record (missing the part of your stack between its actual bottom and what is in `StackBottomBags`).

**Comment by fingolfin** [comment]
Regarding validation, my point is that e.g. `explicit Int(const GAP_Obj gapObj) : super(gapObj) {}` performs no validation at all. I would recommend to write it more like this:
```
explicit Int(const GAP_Obj gapObj) : super(gapObj) {
  if (!IS_INT(gapObj) throw someException;
}
```
At the same time, I'd remove the `Int::isInt` method, it should always return `true` with the above change.

In `Rat`, you could check `TNUM_OBJ(gapObj) == T_RAT || IS_INT(gapObj)`.

And so on.


I just noticed that you also define
```
inline int Int::size() const noexcept
{
  return SIZE_INT(gapObj);
}
```
This will produce another crash or garbage when invoked on a small integer, e.g. try `Int(0).size()`. If you validated `gapObj` to be an integer object, then you can call `GAP_SizeInt`, though, which works correctly for small integers, too.

**Comment by fingolfin** [comment]
More functions that you are using and for which I could produce either a crash or incorrect results include `EqRat`, `AInvRat`, `AbsRat`, `SignRat`, `ModRat`, `PowRat`. As I said, there's a reason we don't just export these.

**Comment by fingolfin** [comment]
> don't see why someone would want to create a C++ object from a GAP handle directly. how would they even get such a handle?

By calling one of the many functions that produce such a handle, but which you don't wrap (e.g. because they don't really fit into any of your classes) ? Most people (all I know other than you) are using `libgap` to access its higher level functionality, and if you e.g. ask a group objects for its order, the result is an integer, and one would thus like to wrap it into an `Int`, but the function to compute the order of a group certainly does not belong into class `Int`...

**Comment by innodocs** [comment]
1. I read your comment on using the GAP dispatch table methods vs the direct calls. did not get around to do it, but I'll have soon a test version which used `EQ` and co from `libgap`, and will compare execution times. I expect that one more indirection will not add a lot of overhead, and then the methods can all be replaced. so I'll end up w/ a GAP handle in Rat which might be an GAP_Int or a GAP_Rat but it won't matter as then the GAP kernel will always call the right functions.

and from the point of view of the user of the C++ interface, it won't matter either if/when a C++ Rat actually stores a GAP int handle. I can add test/conversion functions to convert a `Gap::Rat` to a `Gap::Int` safely.

As to the `Int::size` method, not sure it's needed in the interface at all. pls understand that this is more of a proof-of-concept, and  the final/finalized version will undergo thorough testing as we move it to production.

2. on the issue of `GAP::Init` vs `GAP_Initialize`

>E.g. if you write code that calls your Init wrapper around GAP_Initialize and then at the same level creates GAP objects on the >stack ...

* the comment in your 'libgap.h' file was to call `GAP_Initialize` as soon as possible to avoid those kind of pitfalls. I was assuming that people would adhere to that rule - in fact our rule would be to make it the first thing in `main`. it could be called even sooner (before 'main') if needed by wrapping it in class constructor and then creating a static object of that (init) type - the problem is that there are no valid `argc`/`argv` at that point, so it would need a two step initialization. maybe it defining a static object of type `GAP_Vars` instead could work - I'd have to check the actual code to see what the various functions are doing in detail. as to including the `GAP_VARS` (aka `GAP_Enter`/`GAP_Leave` in too many places) - this can be deadly for performance - I have some timed results, I'll post them soon(ish).
* if you are concerned that the `inline` `Gap::Init` will actually be translated to an actual call and thus not mark the bottom of the stack accurately but one frame up. ok, we can get rid of the `Gap::Init` and call `GAP_Initialize` instead

3. GAP handles and wrapping:
>By calling one of the many functions that produce such a handle, but which you don't wrap (e.g. because they don't really fit >into any of your classes) ?

I've alluded to this earlier: everything will be wrapped. we'll have a small GAP-to-C translator which will be told what to "wrap" when needed. E.g

    sumOfSeries := function(s, e)
     local diff, res;
     if (s > e) then
       return 0;
     fi;
     diff := e - s;
     if diff mod 2 = 0 then
       res := (diff+1) * ((s+e)/2);
     else
       res := ((diff+1)/2) *  (s+e);
     fi;
     return res;
   end;
    
will be translated into

    template<class T>
    T sumOfSeries(T s, T e)  {
      T diff;
     if (s > e)
       return 0;
     diff = e - s;
     return (diff % 2 == 0)
        ?  (diff+1)    * ((s+e)/2)
        : ((diff+1)/2) *  (s+e);
    }

aso asf. library functions in the GAP kernel will also be wrapped (the ones which are needed). so nothing will come into C++ unwrapped.

**Comment by innodocs** [comment]
or on #2 above: just make it a (new) rule that:

- `Gap::Init` is called ASAP in `main`
- no `Gap` functions/functionality is ever used in/called from main

I think that would solve the issue w/ the bottom of the stack.

**Comment by ChrisJefferson** [comment]
Stepping back a little, why would you want a GAP to C++ translator like that? I'll be honest, I strongly suspect you will find it won't work in practice, for non-trivial GAP code. If it did work, what do you expect the benefits to be?

**Comment by innodocs** [comment]
>Stepping back a little, why would you want a GAP to C++ translator like that? I'll be honest, I strongly suspect you will find it >won't work in practice, for non-trivial GAP code. If it did work, what do you expect the benefits to be?

GAP to C++ translator goal: as a first iteration, just wrapping GAP objects for safe use from C++. see template example above (which btw was generated by the translator I was mentioning).

I explained our immediate goal here: for a project we need some math code (most of which exists in one form or the other in GAP - whether kernel C-functions or GAP library functions), and we need to integrate this into an already existing code base (a few mill LOC of C++).

whatever GAP functionality which comes out of the project (not the project itself, obv) we're making available under GPL.

**Comment by fingolfin** [comment]
Regarding GC issues: I am afraid that "`Gap::Init` is called ASAP in `main`" is still not sufficient for correctness, as it still can't determine the correct stack bottom this way. As a workaround you can insert `GAP_Enter(); GAP_Leave();` immediately after `Gap::Init` (strictly speaking that this works relies on an undocumented implementation detail, but I think it is unlikely to change). Note that even this is in general insufficient if the `main` function has local (stack based) variables containing references to GAP objects. To be really safe, your main function should look something like this:
```C
int main(int argc, char **argv)
{
    Gap::Init(argc, argc);
    GAP_Enter(); GAP_Leave();  // trick to make sure the GAP GC knows the correct stack bottom
    return realmain(argc, argv);
}
```
Regarding the rest: of course we understand that your code is work in progress, incomplete, etc. etc. -- our remarks were not meant negative, but rather we were trying to point out perceived flaws, and to also explain why it is unlikely that we'll apply any of the changes your library currently requires. I hope that got clear, and I hope that our free advice was of some use to you.

Anyway, it was quite interesting to see your code, thank you for letting us know about it! However, I don't think it makes sense from our side to keep this issue open; we have no use ourselves for your work, and I don't think most of our users will, at least in its current form, which seems tuned for your specialized usecase -- which of course is fine and makes total sense! It's just that e.g. most users of libgap *do* want to be able to call arbitrary GAP code, e.g. via `GAP_EvalString` or `GAP_CallFuncList` or `GAP_CallFuncArray`.

If you have questions about how things work, I recommend joining our Slack ([here's an invite link](https://join.slack.com/t/gap-system/shared_invite/zt-oc2gxxya-A4_t~sfwtWEHmxqumSYWEQ)) or asking on our mailing list. As was mentioned by @ChrisJefferson, we've also been working on improving many things "under the hood" for better/easier integration over the past years, it's just slow going by nasty and delicate kernel details which take time to be dealt with "properly" by a group of volunteers.

There are a few more open issues related to libgap, some of them may be relevant to you, [see this GitHub issues query](https://github.com/gap-system/gap/issues?q=is%3Aopen+is%3Aissue+label%3A%22topic%3A+libgap%22)

**Comment by innodocs** [comment]
thank you again for taking the time to bring issues to my attention. I will try the Slack invite, thanks for it. I think I attempted to join your mailing list some 10 days ago but never got an acceptance reply from your system.

**Comment by fingolfin** [comment]
Which list did you try to join? It may have gotten stuck in moderation. I can take a look (or try again and ping me if it doesn't work)

**Comment by innodocs** [comment]
I think it was https://mail.gap-system.org/mailman/listinfo/gap

I ran a google search for libgap, which got me to one of your posts

        https://mail.gap-system.org/pipermail/gap/2020-March/001186.html

and then clicked on the _More information about the Gap mailing list_ link at the bottom. Then on that page I went to the 'Subscribing to Gap' section and inserted my info (o...@innodocs.com).

did not receive a confirmation email.
